### PR TITLE
Dynamic Dashboard: Coupons Card UI

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Part of: #12716

## Description

Adds a first iteration of UI related code for the "Most Active Coupons" dashboard card. 

Changes
- View and View model for  "Most Active Coupons" dashboard card.
- Way to store and retrieve selected time range to App settings.
- Open the coupons list screen.
- Update coupon summary text format to match new design.

## Testing instructions

- Enable the feature flag `dynamicDashboardM2` and build the app.
- Confirm that the dashboard still has the same layout as before.
- Tap Edit button and confirm that there's an option for the most active coupons card and it's disabled.
- Enable coupons card and save the change.
- Confirm that the coupons card is displayed on the dashboard.
- Confirm that the UI matches the Figma design.
- Tap View all coupons, confirm that coupons list screen is displayed.
- Tap the calendar icon and change time range, validate that time range selection works.
- Tap the ellipsis menu on the coupons card, confirm that there's an option Hide Coupons. 
- Tap Hide coupons, confirm that the customize screen is displayed.

Disable the `dynamicDashboardM2` flag and confirm that the Coupons card is not available.

## Screenshots

| Light | Dark |
|--------|--------|
| Simulator Screenshot - iPhone 15 Pro Max - 2024-05-14 at 17 10 13 | Simulator Screenshot - iPhone 15 Pro Max - 2024-05-14 at 17 10 39 | 

Time range selection and loading 
Simulator Screen Recording - iPhone 15 Pro Max - 2024-05-14 at 17 11 57

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Yosemite/Yosemite/Stores/AppSettingsStore.swift [hunk 1]
```diff
@@ -1020,6 +1024,19 @@ private extension AppSettingsStore {
...
+        let updatedSettings = storeSettings.copy(lastSelectedMostActiveCouponsTimeRange: timeRange.rawValue)
+        setStoreSettings(settings: updatedSettings, for: siteID)
+    }
+
+    func loadLastSelectedMostActiveCouponsTimeRange(siteID: Int64, onCompletion: (MostActiveCouponsTimeRange?) -> Void) {
+        let storeSettings = getStoreSettings(for: siteID)
+        let timeRangeRawValue = storeSettings.lastSelectedMostActiveCouponsTimeRange
+        let timeRange = MostActiveCouponsTimeRange(rawValue: timeRangeRawValue)
+        onCompletion(timeRange)
+    }
```
Comment (selanthiraiyan) [inline Yosemite/Yosemite/Stores/AppSettingsStore.swift:1039]: Unit tests will be added as part of https://github.com/woocommerce/woocommerce-ios/issues/12728

**Comment by selanthiraiyan** [comment]
Marking as a draft to remove the "All time" option as there is no way to fetch all-time info from API.

**Comment by selanthiraiyan** [comment]
Closing this PR. I will raise a new one to have a clear commit history.

**Pull request closed without merging**